etcdwatcher/etcdwatcher.go

```diff
@@ -0,0 +1,329 @@
+	Watch(ctx context.Context, key string, opts ...clientv3.OpOption) clientv3.WatchChan
+}
+
+// EventType represents the type of event that occurred.
+type EventType string
+
+const (
+	// PutEvent represents a key creation or update event.
+	PutEvent EventType = "PUT"
+	// DeleteEvent represents a key deletion event.
+	DeleteEvent EventType = "DELETE"
+)
+
+// WatchMode defines how the watcher handles events.
+type WatchMode string
+
+const (
+	// AllMode emits all events including updates for keys that have already been emitted.
+	AllMode WatchMode = "ALL"
+	// OnceMode Emit each key at most once across the entire lifecycle (initial load + watch).
+	// Subsequent events for that key are ignored, regardless of type.
+	// this mode can only be used with prefix watches.
+	OnceMode WatchMode = "ONCE"
+	// NoneMode will only emit the existing keys at the time of start and then stop.
+	NoneMode WatchMode = "NONE"
+)
+
+// Event represents a single etcd key event.
+type Event[T any] struct {
+	// Key is the etcd key that changed.
+	Key string
+	// Type is the type of event (PUT or DELETE).
+	Type EventType
+	// Value is the unmarshalled value of the key (for PUT events).
+	Value T
+	// Revision is the etcd revision of the event.
+	Revision int64
+}
+
+// WatchEventType defines what types of events to watch for.
+type WatchEventType string
+
+const (
+	// PutWatchEventType watches for put events only.
+	PutWatchEventType WatchEventType = "PUT"
+	// DeleteWatchEventType watches for delete events only.
+	DeleteWatchEventType WatchEventType = "DELETE"
+	// AllWatchEventType watches for both put and delete events.
+	AllWatchEventType WatchEventType = "ALL"
+)
+
+// Config holds the configuration for the watcher.
+type Config[T any] struct {
+	// EventTypes specifies which event types to watch for.
+	EventTypes WatchEventType
+	// Mode specifies the watch mode.
+	Mode WatchMode
+	// Filter is an optional function to filter events.
+	Filter func(Event[T]) bool
+}
+
+// Watcher provides methods for watching etcd keys.
+type Watcher[T any] struct {
+	client   etcdClient
+	config   Config[T]
+	key      string
+	isPrefix bool
+}
+
+// New creates a new Watcher instance.
+func New[T any](client etcdClient, key string, isPrefix bool, config Config[T]) (*Watcher[T], error) {
+	// Set defaults
+	if config.EventTypes == "" {
+		config.EventTypes = AllWatchEventType
+	}
+	if config.Mode == "" {
+		config.Mode = AllMode
+	}
+
+	if config.Mode == OnceMode && !isPrefix {
+		return nil, fmt.Errorf("once mode can only be used with prefix watches")
+	}
+
+	return &Watcher[T]{
+		client:   client,
+		config:   config,
+		key:      key,
+		isPrefix: isPrefix,
+	}, nil
+}
+
+// LoadAndWatch loads initial data and then starts watching for changes.
+// It returns the initial events, a channel for subsequent events and a function to stop watching for events.
+func (w *Watcher[T]) LoadAndWatch(ctx context.Context) ([]Event[T], <-chan Event[T], <-chan error, func()) {
```

Comment:
We only need 1 channel and 1 leave function. 
```suggestion
func (w *Watcher[T]) LoadAndWatch(ctx context.Context) (initialLoad []Event[T], initialLoadErr error, watchCh <-chan EventOrError[T], leave func()) {
```